aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1335,55 +1372,32 @@ static inline Tensor& bmm_out_or_baddbmm_(Tensor& self_or_result, const Tensor&
       }
     }
   }
-  return self_or_result;
-}
-
-
-Tensor baddbmm_cpu(const Tensor& self, const Tensor& batch1, const Tensor& batch2, const Scalar& beta, const Scalar& alpha) {
-  Tensor result = at::empty({0}, self.options());
-  return at::native::baddbmm_out_cpu(self, batch1, batch2, beta, alpha, result);
+  return;
 }
 
-Tensor& baddbmm_out_cpu(const Tensor& self_, const Tensor& batch1, const Tensor& batch2, const Scalar& beta, const Scalar& alpha, Tensor &result) {
-  auto self = expand_size(self_, {batch1.size(0), batch1.size(1), batch2.size(2)}, "baddbmm");
-  result.resize_(self->sizes());
-  result.copy_(*self);
```

Comment:
I feel like we are losing this copy. But, somehow all the tests are passing.

Reply:
I thought about it before, and I gave it a thought again.

My observation is, `expand_size(self_, {batch1.size(0), batch1.size(1), batch2.size(2)})` calls `self_.expand` method (if required).

Now the thing is, if `result` tensor's size is already (or if we set explicitly using `set_output`): `{batch1.size(0), batch1.size(1), batch2.size(2)}` - and `self` is not of the required size, then `copy_` will resize the `self` tensor to `result` tensor's size.
 
```python
In [17]: M = torch.randn(10, 1, 5)
In [18]: _out = torch.empty((10, 3, 5))

In [20]: _out.copy_(M).size()
Out[20]: torch.Size([10, 3, 5])
```

And we do this `copy_` in the impl function now (after your suggestion in the review comment above, earlier it was in the meta function) if `result` is not same as `self`.

Now, `result` will be same as `self` if:

1. It's an in-place call (auto-generated code sets `result` as `self` in the meta function itself).
2. `out` is `self` (like: `torch.baddbmm(M, batch1, batch2, out=M)`). -- **I'm not resizing for this case, and that's where it's wrong, please see below**

On upstream:

```python
In [43]: M = torch.randn(10, 1, 5)

In [44]: torch.baddbmm(M, batch1, batch2, out=M).size()
Out[44]: torch.Size([10, 3, 5])
```

On this branch:

```python
In [8]: M = torch.randn(10, 1, 5)

In [9]: torch.baddbmm(M, batch1, batch2, out=M).size()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-9-4d21e89d2e08> in <module>
----> 1 torch.baddbmm(M, batch1, batch2, out=M).size()

RuntimeError: Expected self_sizes[0] == bs && self_sizes[1] == res_rows && self_sizes[2] == res_cols to be true, but got false.  (Could this error message be improved?  If so, please report an enhancement request to PyTorch.)
```

We need to test this behavior as well in `test_torch.py`. Alright, I'll fix this now.

Thanks for asking this question, @ysiraichi - sorry for the long message. ;)

Reply:
I'm stuck here:

How would you differentiate these 2 calls (see below) from meta function? (Note: please see the comment above on why we need to differentiate b/w these calls)

```python
>>> M.baddbmm_(batch1, batch2)  # [1] In-Place
>>> torch.baddbmm(M, batch1, batch2, out=M)  # [2] out=self tensor
```

From the auto-generated code for in-place variant in `build/aten/src/ATen/RegisterCPU.cpp`:

```cpp
struct structured_baddbmm_out_cpu_inplace final : public at::native::structured_baddbmm_out_cpu {
     structured_baddbmm_out_cpu_inplace(Tensor& self) : outputs_{std::ref(self)} {}
     // ...
 };
```

As `outputs_` is initialized with `std::ref(self)` so `result.is_same(self)` is always gonna be `true`.

cc: @ezyang @bdhirsh @ysiraichi

Similarly for [2] `out=M` call, `result.is_same(self)` will be `true`. How should we differentiate b/w these calls?

Reply:
Hmm, this actually seems more i-line with our out= behavior: for most other out= ops, if you pass in an output tensor with an incorrect size we raise an error instead of transparently resizing the output for you (unless the tensor has zero size).

I'm just not sure if we want a deprecation cycle, since it's not clear to me how commonly this behavior is relied on (if I had to guess though, it seems like a pretty uncommon use case?). I also don't think you have an easy way with structured kernels to differentiate those two cases.

Reply:
Now, I encountered this behaviour in a test in `test/test_linalg.py` the other day:
https://github.com/pytorch/pytorch/blob/88032d894311e5c0aed8bbc21a4306bc6be4af82/test/test_linalg.py#L4978
Note that the shape is not being passed correctly:
https://github.com/pytorch/pytorch/blob/88032d894311e5c0aed8bbc21a4306bc6be4af82/test/test_linalg.py#L4988-L4995

This will be fixed in https://github.com/pytorch/pytorch/pull/64387 , but it signals that there may be other places where this behaviour is used.

Note that, at the moment, if an incorrectly shaped out is used, a warning is issued. As such, I think that just making that warning into an error may be a reasonable behaviour here.

Reply:
Thanks for the valuable inputs, @Lezcano, and @bdhirsh. Following the suggestions, we now raise an error for incorrectly shaped `out` (earlier we allowed resizing). However, this behavior is only for `baddbmm`. We resize the output for `bmm`, there is a test that uses incorrect shapes for `out` here: https://github.com/pytorch/pytorch/blob/master/test/test_torch.py#L8394.

The only thing that concerns me is the error messages for `inplace` call and `out=M` call. The error with in-place call should have `self_sizes` mentioned instead of `result_sizes`, but that will need distinction b/w the 2 calls.

```python
>>> M = torch.randn(10, 1, 5)  # Incorrect shape
>>> batch1 = torch.randn(10, 3, 4)
>>> batch2 = torch.randn(10, 4, 5)

>>> M.baddbmm_(batch1, batch2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Expected result_sizes[0] == bs && result_sizes[1] == res_rows && result_sizes[2] == res_cols to be true, but got false.  (Could this error message be improved?  If so, please report an enhancement request to PyTorch.)

>>> torch.baddbmm(M, batch1, batch2, out=M)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: Expected result_sizes[0] == bs && result_sizes[1] == res_rows && result_sizes[2] == res_cols to be true, but got false.  (Could this error message be improved?  If so, please report an enhancement request to PyTorch.)
```

The tests might need another look, which I personally feel can be done in a follow-up PR.

Reply:
> How would you differentiate these 2 calls (see below) from meta function? (Note: please see the comment above on why we need to differentiate b/w these calls)

While I agree with @bdhirsh (i.e. you can't), you can resize the output tensors conditionally by using `set_output`. So, instead of writing:

```c++
auto& result = meta.maybe_get_output(0);
if (is_bmm || !result.defined()) {
  meta.set_output({bs, res_rows, res_cols}, batch1.options());
} else {
  const auto result_sizes = result.sizes();
  TORCH_CHECK(result_sizes[0] == bs && result_sizes[1] == res_rows && result_sizes[2] == res_cols);
}
```

You could write:

```c++
// 'set_output' does not resize for in-place calls
meta.set_output({bs, res_rows, res_cols}, batch1.options());
// Error is raised if called from the in-place overload with incorrect self shape
TORCH_CHECK(result_sizes[0] == bs && result_sizes[1] == res_rows && result_sizes[2] == res_cols);
```

This works because the in-place implementation of `set_output` does no resizing!

Reply:
@ysiraichi good catch, that seems like it should work (and also raise the warning). It's probably worth quickly confirming that with that code change, we now get the deprecation warning:
```
>>> M = torch.randn(10, 1, 5)  # Incorrect shape
>>> batch1 = torch.randn(10, 3, 4)
>>> batch2 = torch.randn(10, 4, 5)

>>> torch.baddbmm(M, batch1, batch2, out=M)
# (we should see a deprecation warning, because meta.set_output calls at::native::set_output)

Reply:
This is really cool, @ysiraichi! I wasn't aware that `meta.set_output()` doesn't resize for in-place calls. This helps!

Here is how the use-cases look like with the revised changes:

```python
In [11]: M = torch.randn(10, 1, 5)
In [12]: batch1 = torch.randn(10, 3, 4)
In [13]: batch2 = torch.randn(10, 4, 5)

In [14]: torch.baddbmm(M, batch1, batch2).size()
Out[14]: torch.Size([10, 3, 5])

In [15]: M.baddbmm_(batch1, batch2).size()
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-15-ffc510337fae> in <module>
----> 1 M.baddbmm_(batch1, batch2).size()

RuntimeError: Expected result_sizes[0] == bs && result_sizes[1] == res_rows && result_sizes[2] == res_cols to be true, but got false.  (Could this error message be improved?  If so, please report an enhancement request to PyTorch.)

In [16]: torch.baddbmm(M, batch1, batch2, out=M).size()
<ipython-input-16-4d21e89d2e08>:1: UserWarning: An output with one or more elements was resized since it had shape [10, 1, 5], which does not match the required output shape [10, 3, 5].This behavior is deprecated, and in a future PyTorch release outputs will not be resized unless they have zero elements. You can explicitly reuse an out tensor t by resizing it, inplace, to zero elements with t.resize_(0). (Triggered internally at  ../aten/src/ATen/native/Resize.cpp:16.)
  torch.baddbmm(M, batch1, batch2, out=M).size()
Out[16]: torch.Size([10, 3, 5])
```

We do see a deprecation warning with `out=M`, and we also get an error for the in-place call. I've made the required commits, the tests should pass and this PR should hopefully look good now. Welcoming reviews now. :)